ext/drutil/drutil.c

```diff
@@ -58,12 +58,44 @@
 /* for inserting an app instruction, which must have a translation ("xl8") field */
 #define PREXL8 instrlist_preinsert
 
+#ifdef X86
+static uint drutil_xsave_area_size;
+#endif
+
 /***************************************************************************
  * INIT
  */
 
 static int drutil_init_count;
 
+#ifdef X86
+
+static inline void
+native_cpuid(uint *eax, uint *ebx, uint *ecx, uint *edx)
+{
+#    ifdef UNIX
+    /* We need to do this ebx trick, because ebx might be used for fPIC,
+     * and gcc < 5 chokes on it. This can get removed and replaced by
+     * a "=b" constraint when moving to gcc-5.
+     */
+    asm volatile("xchgl\t%%ebx, %k1\n\t"
+                 "cpuid\n\t"
+                 "xchgl\t%%ebx, %k1\n\t"
+                 : "=a"(*eax), "=&r"(*ebx), "=c"(*ecx), "=d"(*edx)
+                 : "0"(*eax), "2"(*ecx));
+#    endif
+}
+
+static inline void
+cpuid(uint op, uint subop, uint *eax, uint *ebx, uint *ecx, uint *edx)
+{
+    *eax = op;
+    *ecx = subop;
+    native_cpuid(eax, ebx, ecx, edx);
```

Comment:
IMHO this function should be where the Windows-Linux split should exist, providing a single interface for both, rather than pushing that complexity further out into the high-level code like drutil_init.  Splitting here would simplify and shrink the amount of code needed; make drutil_init shorter, cleaner, and cross-platform; make it easier to change this code to use a shared library routine for cpuid in the future; etc.

Reply:
Ok. Done.